src/timemachine.c

```diff
@@ -0,0 +1,613 @@
+#include "debug.h"
+#include "decode.h"
+#include "detect.h"
+#include "conf.h"
+#include "output.h"
+
+#include "threads.h"
+#include "threadvars.h"
+#include "tm-threads.h"
+
+#include "util-error.h"
+#include "util-debug.h"
+#include "util-time.h"
+#include "util-byte.h"
+#include "util-misc.h"
+#include "util-cpu.h"
+#include "util-atomic.h"
+#include "util-time.h"
+
+#include "timemachine.h"
+#include "timemachine-flow.h"
+#include "timemachine-heap.h"
+#include "timemachine-output.h"
+#include "timemachine-packet.h"
+
+#define MODULE_NAME                               "TimeMachine"
+#define DEFAULT_MAX_MEMORY                        1024 * 1024 * 1024
+#define DEFAULT_MAX_PACKET_SIZE                   1500
+#define DEFAULT_HEAP_PREALLOC_COUNT               5000
+#define DEFAULT_HEAP_EXPAND_BY                    1000
+#define DEFAULT_OUTPUT_TIMEOUT                    21600
+
+SC_ATOMIC_DECLARE(uint32_t, thread_cnt);
+
+/* global pcap data for when we're using multi mode. At exit we'll
+ * merge counters into this one and then report counters. */
+static TimeMachineData *g_tm_data = NULL;
+
+static TmEcode TimeMachineProcess(ThreadVars *, Packet *, void *, PacketQueue *, PacketQueue *);
+static TmEcode TimeMachineInit(ThreadVars *, void *, void **);
+static TmEcode TimeMachineDeinit(ThreadVars *, void *);
+static OutputCtx *TimeMachineInitCtx(ConfNode *);
+static void TimeMachineDeInitCtx(OutputCtx *);
+
+void TmModuleTimeMachineRegister(void)
+{
+    tmm_modules[TMM_TIMEMACHINE].name = MODULE_NAME;
+    tmm_modules[TMM_TIMEMACHINE].ThreadInit = TimeMachineInit;
+    tmm_modules[TMM_TIMEMACHINE].Func = TimeMachineProcess;
+    tmm_modules[TMM_TIMEMACHINE].ThreadDeinit = TimeMachineDeinit;
+    tmm_modules[TMM_TIMEMACHINE].RegisterTests = NULL;
+
+    OutputRegisterModule(MODULE_NAME, "timemachine", TimeMachineInitCtx);
+
+    SC_ATOMIC_INIT(thread_cnt);
+    return;
+}
+
+/**
+ * \brief TimeMachine main logging function
+ *
+ * \param t threadvar
+ * \param p packet
+ * \param data thread module specific data
+ * \param pq pre-packet-queue
+ * \param postpq post-packet-queue
+ *
+ * \retval TM_ECODE_OK on succes
+ * \retval TM_ECODE_FAILED on serious error
+ */
+static TmEcode TimeMachineProcess(ThreadVars *t, Packet *p, void *thread_data, PacketQueue *pq,
+                 PacketQueue *postpq)
+{    
+    TimeMachineFlow *flow;
+    TimeMachineHeap *heap;
+    TimeMachineMemPool *mem_pool;
+    TimeMachinePacket *packet;
+    struct pcap_pkthdr pkthdr;
+    struct timeval current_time;
+    
+    TimeGet(&current_time);
+    
+    TimeMachineThreadData *td = (TimeMachineThreadData *)thread_data;
+    TimeMachineData *tm = td->tm_data;
+    
+    /* make sure packet has real data */
+    if (GET_PKT_PAYLOAD_LEN(p) == 0) {
+        return TM_ECODE_OK;
+    }
+    
+    /* check whether a corresponding flow already exists */
+    if (!(flow = TimeMachineFlowLookup(td->flows, p))) {
+        flow = TimeMachineFlowNew(td->flows, p);       
```

Comment:
TimeMachineFlowNew will exit instead of return NULL on memory allocation failure. Exits should not occur in the packet path.  Instead, TimeMachineProcess should return TM_ECODE_FAILED.